identities/models.py

```diff
@@ -44,3 +48,74 @@ class Identity(models.Model):
 
     def __str__(self):
         return str(self.id)
+
+    def remove_details(self, user):
+        updated_details = {}
+        for attribute, value in self.details.iteritems():
+            updated_details[attribute] = "removed"
+        self.details = updated_details
+        self.updated_at = datetime.datetime.now()
+        self.updated_by = user
+        self.save()
+
+
+class OptOut(models.Model):
+    """An optout"""
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False,
+                          help_text="UUID this optout request.")
+    identity = models.ForeignKey(Identity, null=True,
+                                 help_text="UUID for the identity opting out.")
+    address_type = models.CharField(
+        null=False, max_length=50, default="",
+        help_text="Address type used to identify the identity.")
+    address = models.CharField(
+        null=False, max_length=255, default="",
+        help_text="Address used to identify the identity.")
+    request_source = models.CharField(
+        null=False, max_length=100,
+        help_text="Service that the optout was requested from.")
+    requestor_source_id = models.CharField(
+        null=True, max_length=100,
+        help_text="ID for the user requesting the optout on the service that\
+        it was requested from.")
+    created_at = models.DateTimeField(auto_now_add=True,
+                                      help_text="Time request was received.")
+    created_by = models.ForeignKey(User, related_name='optout_created',
+                                   null=True,
+                                   help_text="User creating the OptOut")
+
+    user = property(lambda self: self.created_by)
+
+
+@receiver(post_save, sender=OptOut)
+def handle_optout(sender, instance, **kwargs):
+    if instance.identity is not None:
+        identity = instance.identity
+    else:
+        filter_string = \
+            "details__addresses__" + instance.address_type + "__has_key"
+        filter_value = instance.address
+        identities = Identity.objects.filter(**{filter_string: filter_value})
+        try:
+            identity = identities[0]
```

Comment:
Worried about grabbing the first identity here -- I can't see how guessing which identity to use can be the right thing to do for optouts.